File: java-commons/concurrent/src/main/java/cn/wjybxx/concurrent/AggregateOptions.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.concurrent;

import cn.wjybxx.base.annotation.Internal;

/**
 * future的聚合选项
 *
 * @author wjybxx
 * date 2023/4/12
 */
@Internal
public final class AggregateOptions {

    private final boolean anyOf;
    public final int successRequire;
    public final boolean failFast;

    AggregateOptions(boolean anyOf, int successRequire, boolean failFast) {
        this.anyOf = anyOf;
        this.successRequire = successRequire;
        this.failFast = failFast;
    }

    public boolean isAnyOf() {
        return anyOf;
    }

    private static final AggregateOptions ANY = new AggregateOptions(true, 0, false);

    /** 任意一个完成 */
    public static AggregateOptions anyOf() {
        return ANY;
    }

    /**
     * 成功完成n个
     *
     * @param successRequire 需要成功完成的数量
     * @param failFast       是否快速失败
     */
    public static AggregateOptions selectN(int successRequire, boolean failFast) {
        if (successRequire < 0) {
            throw new IllegalArgumentException("successRequire < 0");
        }
        return new AggregateOptions(false, successRequire, failFast);
    }
}

File: csharp/Wjybxx.Commons.Concurrent/src/AggregateOptions.cs
#region LICENSE

// Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;

namespace Wjybxx.Commons.Concurrent
{
internal readonly struct AggregateOptions
{
    private readonly byte type;
    public readonly int required;
    public readonly bool failFast;

    private AggregateOptions(byte type, int required, bool failFast) {
        this.type = type;
        this.required = required;
        this.failFast = failFast;
    }

    public bool IsWhenAny => type == TYPE_ANY;
    public bool IsWhenAll => type == TYPE_WHEN_ALL;

    private const byte TYPE_ANY = 0;
    private const byte TYPE_WHEN_ALL = 1;
    private const byte TYPE_SELECT_MANY = 2;

    private static readonly AggregateOptions WHEN_ANY = new AggregateOptions(TYPE_ANY, 0, false);
    private static readonly AggregateOptions WHEN_ALL = new AggregateOptions(TYPE_WHEN_ALL, 0, false);

    /// <summary>
    /// 任意一个完成
    /// </summary>
    /// <returns></returns>
    public static AggregateOptions WhenAny() {
        return WHEN_ANY;
    }

    /// <summary>
    /// 所有任务完成（无快速失败逻辑）
    /// </summary>
    /// <returns></returns>
    public static AggregateOptions WhenAll() {
        return WHEN_ALL;
    }

    /// <summary>
    /// 成功完成n个
    /// </summary>
    /// <param name="required">需要成功完成的数量</param>
    /// <param name="failFast">是否快速失败</param>
    /// <returns></returns>
    /// <exception cref="ArgumentException"></exception>
    public static AggregateOptions SelectN(int required, bool failFast) {
        if (required < 0) {
            throw new ArgumentException("required cannot be negative");
        }
        return new AggregateOptions(TYPE_SELECT_MANY, required, failFast);
    }
}
}